lkl: macvtap support for virtio-net backend

thehajime opened the pull request:
This new virtio_net backend is used as usual tun/tap device, but doesn't
require dedicated bridge device.  Instead the parent netdev is used as a
bridge device so that no additional interface configuration is needed
for LKL.

This is similar to raw socket (af_packet) virtio_net backend  but the
macvtap backend doesn't need root privilege (or a capability) during an
execution of LKL kernel.

This commit also implements offload feature as tap backend currently has
(TSO4, CSUM).

In order to use this, one needs to properly configure the host
netdevice, like

 % sudo ip link add link eth0 name vtap0 type macvtap mode passthru
 % sudo ip link set dev vtap0 up
 % sudo chown thehajime /dev/tap22

then specifies the new LKL_HIJACK_NET_IFTYPE, macvtap, with the
parameter LKL_HIJACK_NET_IFPARAMS, by /dev/tap22, which depends on your
executed environment.

LKL_HIJACK_NET_IFTYPE=macvtap LKL_HIJACK_NET_IFPARAMS=/dev/tap22
lkl-hijack.sh (application)

Signed-off-by: Hajime Tazaki thehajime@gmail.com

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lkl/linux/186)

<!-- Reviewable:end -->

opurdila commented:
Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 31 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOGaV98Z7wlKI7g65e7:-KOGaV98Z7wlKI7g65e8:782438594) ([raw file](https://github.com/lkl/linux/blob/bebdf30ecffd75e4c08356ecfbf063b52e851c63/tools/lkl/lib/virtio_net_macvtap.c#L31)):*
> ```C
> #define BIT(x) (1ULL << x)
> 
> struct lkl_netdev *lkl_netdev_macvtap_create(const char *path, int offload)
> ```

This looks very similar with lkl_netdev_tap_create(). I think its worth creating a helper function that both lkl_netdev_tap_create and lkl_netdev_macvtap_create can wrap. Or, we can add a new parameter to lkl_netdev_tap_create?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

liuyuan10 commented:
Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 31 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOGaV98Z7wlKI7g65e7:-KOGjGQr818roia8oxgv:5172523) ([raw file](https://github.com/lkl/linux/blob/bebdf30ecffd75e4c08356ecfbf063b52e851c63/tools/lkl/lib/virtio_net_macvtap.c#L31)):*
<details><summary><i>Previously, opurdila (Octavian Purdila) wrote…</i></summary>
> This looks very similar with lkl_netdev_tap_create(). I think its worth creating a helper function that both lkl_netdev_tap_create and lkl_netdev_macvtap_create can wrap. Or, we can add a new parameter to lkl_netdev_tap_create?
</details>
+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

thehajime commented:
Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 31 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOGaV98Z7wlKI7g65e7:-KOII8c5_LY57r8aC6PV:1095092837) ([raw file](https://github.com/lkl/linux/blob/bebdf30ecffd75e4c08356ecfbf063b52e851c63/tools/lkl/lib/virtio_net_macvtap.c#L31)):*
<details><summary><i>Previously, liuyuan10 (Yuan Liu) wrote…</i></summary>
> +1
</details>
I've introduced `lkl_netdev_tap_init()` to wrap the shared part of both.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

thehajime commented:
btw, circleci infrastructure has some issue so it doesn't report test has passed.

opurdila commented:
Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 31 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOGaV98Z7wlKI7g65e7:-KOJdZFPM5-Fhgt3EwFs:-783801356) ([raw file](https://github.com/lkl/linux/blob/bebdf30ecffd75e4c08356ecfbf063b52e851c63/tools/lkl/lib/virtio_net_macvtap.c#L31)):*
<details><summary><i>Previously, thehajime (Hajime Tazaki) wrote…</i></summary>
> I've introduced `lkl_netdev_tap_init()` to wrap the shared part of both.
</details>
Since we only have one new function I think it would be better to keep it in virtio_net_tap.c and not create a new file. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

thehajime commented:
Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 31 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOGaV98Z7wlKI7g65e7:-KOK855QVAhRWy_bjpar:-201507078) ([raw file](https://github.com/lkl/linux/blob/bebdf30ecffd75e4c08356ecfbf063b52e851c63/tools/lkl/lib/virtio_net_macvtap.c#L31)):*
<details><summary><i>Previously, opurdila (Octavian Purdila) wrote…</i></summary>
> Since we only have one new function I think it would be better to keep it in virtio_net_tap.c and not create a new file. 
</details>
I wanted to have new file because

1. we will have additional macvtap specific configuration when the mode configuration is different from passthrough (https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Virtualization_Administration_Guide/sect-attch-nic-physdev.html).  For instance, if the mode is vepa, it needs to configure MAC address in LKL side.
2. since other backends (tap/raw/dpdk/vde) have also independent files, having new file is straightforward and easier when someone starts to look at code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

thehajime commented:
rebased and fixed the test failure.

liuyuan10 commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[tools/lkl/lib/virtio_net_linux_fdnet.h, line 5 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLKklG4noj9LuDCXFh:-KOLKklH-nQ8YrkJiORx:694559942) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_linux_fdnet.h#L5)):*
> ```objc
> 
> enum tap_type {
> 	LKL_TUNTAP,
> ```

I think TUN and TAP are two different things. People can also have an device implementation using TUN?

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 18 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLJr1sIBCAqC0YXmi1:-KOLJr1sIBCAqC0YXmi2:-2063956945) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_macvtap.c#L18)):*
> ```C
>  */
> 
> #include <stdio.h>
> ```

Does it need this many headers?

---

*[tools/lkl/lib/virtio_net_tap.c, line 92 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLKXLMzYFRTFryjyfk:-KOLKXLMzYFRTFryjyfl:6064850) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_tap.c#L92)):*
> ```C
> 	strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
> 
> 	return (struct lkl_netdev *)lkl_netdev_tap_init(LKL_MACVTAP, path,
> ```

LKL_TUNTAP?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

thehajime commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[tools/lkl/lib/virtio_net_linux_fdnet.h, line 5 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLKklG4noj9LuDCXFh:-KOMqSfd7_Xe5Vv8ohiT:505107846) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_linux_fdnet.h#L5)):*
<details><summary><i>Previously, liuyuan10 (Yuan Liu) wrote…</i></summary>
> I think TUN and TAP are two different things. People can also have an device implementation using TUN?
</details>
indeed.  using LKL_TAP should be appropriate.
I deleted this enum since it's not used at all.

---

*[tools/lkl/lib/virtio_net_macvtap.c, line 18 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLJr1sIBCAqC0YXmi1:-KOMqf6Cjc3BN2b3EAqe:-1620474753) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_macvtap.c#L18)):*
<details><summary><i>Previously, liuyuan10 (Yuan Liu) wrote…</i></summary>
> Does it need this many headers?
</details>
fixed.

---

*[tools/lkl/lib/virtio_net_tap.c, line 92 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/186#-KOLKXLMzYFRTFryjyfk:-KOMqgtKC8foIE3zibVI:1145741023) ([raw file](https://github.com/lkl/linux/blob/96cf1c8bb38920b9a1172f091f1e164c745afe29/tools/lkl/lib/virtio_net_tap.c#L92)):*
<details><summary><i>Previously, liuyuan10 (Yuan Liu) wrote…</i></summary>
> LKL_TUNTAP?
</details>
indeed, it should be.  again, I deleted this too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186)*
<!-- Sent from Reviewable.io -->

liuyuan10 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186#-:-KOOEUrBVUd9CbXEMWiN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

opurdila commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r1, 2 of 6 files at r2, 2 of 2 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/186#-:-KOOfZRVSq-Pr2Pt__NN:bnfp4nl)*
<!-- Sent from Reviewable.io -->
